cmd/api/src/api/middleware/compression.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2023 Specter Ops, Inc.
+//
+// Licensed under the Apache License, Version 2.0
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package middleware
+
+import (
+	"compress/gzip"
+	"compress/zlib"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/specterops/bloodhound/headers"
+	"github.com/specterops/bloodhound/log"
+	"github.com/specterops/bloodhound/src/api"
+)
+
+var unsupportedEncodingError = errors.New("content encoding is not supported")
+
+type GzipResponseWriter struct {
+	http.ResponseWriter
+	gw *gzip.Writer
+}
+
+func NewGzipResponseWriter(w http.ResponseWriter) *GzipResponseWriter {
+	return &GzipResponseWriter{
+		ResponseWriter: w,
+		gw:             gzip.NewWriter(w),
+	}
+}
+
+func (s *GzipResponseWriter) Write(p []byte) (int, error) {
+	return s.gw.Write(p)
+}
+
+func (s *GzipResponseWriter) Close() error {
+	return s.gw.Close()
+}
+
+func CompressionMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(responseWriter http.ResponseWriter, request *http.Request) {
+		var (
+			gw  *GzipResponseWriter
+			err error
+		)
+
+		if contentEncodingString := strings.Join(request.Header.Values(headers.ContentEncoding.String()), ","); contentEncodingString != "" { // "Content-Encoding: gzip, deflate; Content-Encoding: br;" = "gzip, deflate, br"
+			for _, encoding := range strings.Split(contentEncodingString, ",") {
+				encoding = strings.TrimSpace(encoding)
+				request.Body, err = wrapBody(encoding, request.Body)
+				if err != nil {
+					errMsg := fmt.Sprintf("failed to create reader for %s encoding: %v", encoding, err)
+					log.Warnf(errMsg)
+					if errors.Is(err, unsupportedEncodingError) {
+						api.WriteErrorResponse(request.Context(), api.BuildErrorResponse(http.StatusUnsupportedMediaType, fmt.Sprintf("Error trying to read request: %s", errMsg), request), responseWriter)
+					} else {
+						api.WriteErrorResponse(request.Context(), api.BuildErrorResponse(http.StatusBadRequest, fmt.Sprintf("Error trying to read request: %s", errMsg), request), responseWriter)
```

Comment:
Shouldn't this actually be a 5xx? A failure to process the request is only a client error if the request was malformed or something. If the issue is with our code or server or something then that's still a server error.

Reply:
This case would indicate a failure to instantiate an instance of the compression lib's `ReadCloser`. We would have to dig into the error to derive if it was because of a client issue or not. I'm inclined to agree with @computator  and say 500 is likely the the more appropriate response code.

Reply:
After inspecting how the different readers are instantiated in the stdlib, all of the errors raised are the result of bad data (empty/unexpected EOF, bad header, etc). This means all the failure cases would be the result of a bad request because bad data was sent. A 500 also signals to the client that it cannot do anything different to fix the request, which in this case would be incorrect.